README.md

![build passing](https://travis-ci.org/draganradu/dot-color-js.svg?branch=master) ![License](https://img.shields.io/npm/l/dot-color) ![maintained](https://img.shields.io/maintenance/yes/2020) ![Vulnerabilities](https://img.shields.io/snyk/vulnerabilities/npm/dot-color) ![Node Version](https://img.shields.io/node/v/dot-color)
# Dot color JS

[Demo](http://fotodex.ro/)

This is a modernized version of [simple-color](https://www.npmjs.com/package/simple-color-converter) it uses modern JavaScript getters and setters and all sorts of JavaScript goodness to make one of the lightest but most feature dense color convertors for JavaScript you can find on NPM.

## Usage
Use [link](https://github.com/draganradu/dot-color-js/blob/master/documentation/dot_color_js_documentation.md) for full documentation.
```javascript

const dotColor = require("dot-color")
const color = new dotColor()

color.cmyk = "cmyk 0 50 60 60"
console.log(color.ral) // {lrv: 5, name: "Red Brown", ral: 8012}

color.rgb = "rgb 10 60 240"
console.log(color.ral) // {lrv: 4, name: "Ultramarine Blue", ral: 5002}

console.log(color.hex3) // "03F"
console.log(color.hex8) // 0A3CF0FF"
console.log(color.rgb) // {b: "240", g: "60", r: "10"}
console.log(color.cmyk) // {c: 96, k: 6, m: 75, y: 0}

```

## Features
1 | [Identify](https://github.com/draganradu/dot-color-js/tree/master/documentation/dot_color_js_documentation.md#2--set-color) color values from string 
2 | [Convert](https://github.com/draganradu/dot-color-js/tree/master/documentation/dot_color_js_documentation.md#4--accepted-colors) to any color format
3 | Generate [color patterns](https://github.com/draganradu/dot-color-js/tree/master/documentation/dot_color_js_documentation.md#8--color-patterns)
4 | Use [logical javascript design pattern](https://github.com/draganradu/dot-color-js/tree/master/documentation/dot_color_js_documentation.md#6--utility-keys) for custom color objects

## Supported colors
The main thing we do better than any other color convertor is the ability to get any string and figure out if there is a color buried in there. We designed this package to be used behind an input field where the values can range from "RED" to "CMYK 30 20 10 5" and even "C 30 K 5 M 20 Y 10". The rule is: if you can read it as a color, we identify it as a valid color.

[Accepted colors input test](exemple_color.md)

| color system      |  object   | array     | string (basic format) |
| ---               |  ---      | ---       | ---    |
| cmyk              | {c: 39, m: 0, y: 39, k: 7} | [39, 0, 39.7] | 'cmyk 39 0 39 7'
| grayscale         |           |           | '78' |
| hex3              |           |           | '#9E9' |
| hex4              |           |           | '#9E9F' |
| hex6              |           |           | '#90EE90' |
| hex8              |           |           | '#90EE90FF' |
| html              |           |           | 'Light Green' |
| hsl               | { h: 120, s: 73.4, l: 74.9 } | [120, 73.4, 74.9] | 'hsl 120 73.4 74.9' |
| hsv               | { h: 120, s: 39.4, l: 93.3 } | [120, 39.4, 93.3] | 'hsl 120 39.4 93.3' |
| lab               |  {l: 86.5, a: -46.3, b: 36.9} | [86.5, -46.3, 36.9] | 'lab 86.5 -46.3 36.9' |
| pantone           | {name: '358C'} | | 'pantone 358C' |
| ral   | { ral : 6019 } | | 'ral 6019'|
| rgb | { r: 144, g: 238, b: 144 } | [144, 238, 144] | 'rgb 144 238 144'|
| rgba | { r: 144, g: 238, b: 144, a: 1 } | [144, 238, 144, 1] | 'rgba 144 238 144 1'|
| rgb decimal | | | 'rgb decimal 6812065' | 
| w | { r: 144, g: 238, b: 144, a: 1 } | [144, 238, 144, 1] | 'w 544' |
| xyz | { x: 44, y: 69, z: 45 } | [44, 69, 45] | 'xyz 44 69 45' |
| yuv | { y: 180, u: 113.2, v: 73.5 } | [ 180, 113.2, 73.5 ] | 'yuv 180 113.2 73.5' |


## Contribute
If there are any features you would like to support, or want to add it directly please send us a pull request. I`m more than happy.
If you have any suggestions or you spotted an aberrant behavior or bugs, don't hesitate to send me an email. 

There is only one rule, no breaking changes

## License
Copyright © 2020, Radu Dragan. Licensed under the [MIT License](https://github.com/draganradu/dot-color-js/blob/master/LICENSE).

## Thank you

I`m Radu, Thank you for using my color convertor, I hope it is useful for you. I'm genuinely excited to build this kind of solutions.

CHANGELOG.md
## 1.1.1

### Testing

* Added Jest for testing

### Code Quality

* Added Change log

### Compare

* Added Compare, an array of colors to the color set, this returns an order array based on closest color.

### Abstract color

* Fixed the Ral String abstract internal Issue.

test/one.test.js


const colorAbstract = require('../dot-color')

const color = new colorAbstract('cmyk 10 20 50 10')
console.log(color.compare('cmyk 60 20 50 10', 'cmyk 20 20 50 10', 'cmyk 1 20 50 10')) 

/*

[ { ral: 1006, name: 'Maize Yellow', lrv: 36 },
  { ral: 1000, name: 'Green Beige', lrv: 50 },
  { ral: 6019, name: 'Pastel Green', lrv: 57 } ]

*/




dot-color.test.js
const dotColor = require('./dot-color')
const testData = require('./test/test.data')
const testApp = new dotColor('')


for(const i of testData) {
    testApp.color = i[0]
    for ( const a in i[1]) {
        const t = (testApp[a] instanceof String)?testApp[a].toString() : testApp[a]
        const d = i[1][a]
        test([a,JSON.stringify(i[0]), JSON.stringify(t)].join(' | '), () => {
            expect(t).toEqual(d)
        })
    }

}

test/test.data.js
module.exports = [
    ['cmyk 10 20 50 6', { format: 'cmyk' }], // cmyk
    ['cMyk 10 20 50 6', { format: 'cmyk' }],
    ['10 20 50 6', { format: 'cmyk' }],
    ['10 20 50 6 cmyk', { format: 'cmyk' }],
    ['c: 10 m:20 y:50 k:6', { format: 'cmyk' }],
    ['c: 10 m:20 k:50 y:6', { format: 'cmyk' }],
    ['c:10m:20k:50y:6', { format: 'cmyk' }],
    ['c10m20k50y6', { format: 'cmyk' }],
    ['cmyk=39,0,39,7', { format: 'cmyk' }],
    ['cmyk(39 0 39 7)', { format: 'cmyk' }],
    ['cmyk(39,0,39,7)', { format: 'cmyk' }],
    ['cmyk(39/0/39/7)', { format: 'cmyk' }],
    ['cmyk(39;0;39;7)', { format: 'cmyk' }],
    ['21', { format: 'grayscale' }], // grayscale
    ['21%', { format: 'grayscale' }],
    ['g 21', { format: 'grayscale' }],
    ['g:21', { format: 'grayscale' }],
    ['g=21', { format: 'grayscale' }],
    ['grayscale 21', { format: 'grayscale' }],
    ['grayscale:21', { format: 'grayscale' }],
    ['grayscale=21', { format: 'grayscale' }],
    ['mono 21', { format: 'grayscale' }],
    ['mono:21', { format: 'grayscale' }],
    ['mono=21', { format: 'grayscale' }],
    ['mono 21%', { format: 'grayscale' }],
    ['9E9', { format: 'hex3' }], // hex 3
    ['#9E9', { format: 'hex3' }],
    ['hex 9E9', { format: 'hex3' }],
    ['hex #9E9', { format: 'hex3' }],
    ['hex3 9E9', { format: 'hex3' }],
    ['hex3 #9E9', { format: 'hex3' }],
    ['hex9E9', { format: 'hex3' }],
    ['hex3 9E9', { format: 'hex3' }],
    ['0x9E9', { format: 'hex3' }],
    ['Ox9E9', { format: 'hex3' }],
    ['hex 0x9E9', { format: 'hex3' }],
    ['hex3 0x9E9', { format: 'hex3' }],
    ['hex3 #0x9E9', { format: 'hex3' }],
    ['hex ox9E9', { format: 'hex3' }],
    ['AF02', {format: 'hex4'}],
    ['FFAA21', {format: 'hex6'}],
    ['1212FF21', {format: 'hex8'}],
    ['magenta', {format: 'html'}],
    ['red', { format: 'html' }],
    ['html red', { format: 'html' }],
    ['hsl 10 20 14', {format: 'hsl'}],
    ['hsv 10 20 14', {format: 'hsv'}],
    ['lab 10 20 14', {format: 'lab'}],
    ['pantone 100', {format: 'pantone'}],
    ['pantone 100C', {format: 'pantone'}],
    ['ral 7015', {format: 'ral'}],
    ['10 20 15', {format: 'rgb'}],
    ['rgb 10 20 15', {format: 'rgb'}],
    ['rgba 10 20 15 0.2', {format: 'rgba'}],
    ['rgb 16711680', { format: 'rgbDecimal' }],
    ['rgbdecimal 16711680', { format: 'rgbDecimal' }],
    ['rgbDecimal 16711680', { format: 'rgbDecimal' }],
    ['rgb decimal 16711680', { format: 'rgbDecimal' }],
    ['rgb numeric 16711680', { format: 'rgbDecimal' }],
    ['w 450', {format: 'w'}],
    ['xyz 10 20 4', {format: 'xyz' }],
    ['yuv 20 20 4', {format: 'yuv'}],
    ['black', { format: 'html' }],
    ['bac', { format: 'hex3' }],
    ['nimic', { format: false }],
    ['c: 10 m:20 k:50 y:6', { format: 'cmyk', cmyk: { c: 10, m: 20, k: 50, y: 6 } }],
    ['g: 30', { format: 'grayscale', cmyk: { c: 0, m: 0, k: 30, y: 0 }}],
    ['rgb 0 255 0', { format: 'rgb', rgb: { r: 0, g: 255, b: 0 }}],
    ['rgb(0 255 0)', { invert: { r: 255, g: 0, b: 255 }}], // invert
    ['rgb 60 8 10', { invert: { r: 8, g: 60, b: 58 }}],
    ['rgb 60 8 10', { invert: { r: 8, g: 60, b: 58 }}],
    ['RAL 1011', { invert: { ral: 5014, name: 'Pigeon Blue', lrv: 19 }}],
    ['rgb 60 8 10', { primary: { r: 255, g: 0, b: 0 }}], // primary
    ['red', { primary: 'Red'}],
    ['RAL 1011', { primary: { ral: 7034, name: 'Yellow Grey', lrv: 25 }}],
    ['ral 7015', { primary: { ral: 8002, name: 'Signal Brown', lrv: 10 }}],
    ['rgb 60 8 10', { secondary: { r: 255, g: 0, b: 255 }}], // secoundary
    ['red', { secondary: 'Yellow'}],
    ['RAL 1011', { secondary: { ral: 1016, name: 'Sulfur Yellow', lrv: 71 }}],
    ['rgb 60 8 10', { tertiary: { r: 255, g: 0, b: 128 }}], // tertiary
    ['red', { tertiary: 'Dark Orange'}],
    ['RAL 1011', { tertiary: { ral: 1017, name: 'Saffron Yellow', lrv: 49 }}],
    ['hsl(10 50 20)', { analogous: [ { h: 340, s: 50, l: 20 }, { h: 10, s: 50, l: 20 }, { h: 40, s: 50, l: 20 } ]}], // analogous
    ['Coral', { analogous: [ 'Salmon', 'Coral', 'Sandy Brown' ]}],
    ['DarkMagenta', { analogous: [ 'Indigo', 'Dark Magenta', 'Brown' ]}],
    ['80', { analogous: [ 80, 80, 80 ]}],
    ['#801', { tints: [ '801', 'A01',  'D01',  'F01',  'F13',  'F45',  'F67',  'F89',  'FBB',  'FDD',  'FFF' ]}], // tints
    ['#801', { shades: [ '000',    '000',    '100',    '200',    '300',    '400',   '500',    '500',    '600',    '700',    '801' ]}], // shades
    ['yellow', { complementary: [ 'Blue', 'Yellow' ]}], // complementary
    ['#801', { complementary: [ '#087', '#801' ]}],
    ['RaL 5001', { complementary: [ 'RAL 8011', 'RAL 5001' ]}],
]


components/identify.js
'use strict'
// const { html, pantone, ral } = require('color_library')
const AcceptedColors = require("./accepted_colors")
const sanitizer = require("./sanitization")
const convertColor = require("./convert_color")

class abstractColor {
    constructor() {
        this.exclusionRegex = {
            cmyk: /hex/g,
            hex: /(?![hxo])[g-z]|([0-9]\w)\s([0-9]\w)\s([0-9]\w)/ig,
            grayscale: /#|pantone/g,
            pantone: /w|rgb|ral/ig,
            rgb: /xyz|lab|hsl|yuv/g,
        }
    }

    abstractMakeInt(a, b = 100) {
        if (typeof a === 'string') {
            let aClean = parseFloat(a.replace(/[^0-9,.]/g, ''))

            if (a.indexOf('%') > -1) {
                return (aClean / 100) * b
            }
            if (parseFloat(aClean) < 1) {
                return aClean
            }
            return aClean
        }

        return isNaN(a) ? false : a
    }

    procentRequire(number1, number2) {
        if (number1 > 1 && number2 > 1) {
            return false
        } else {
            return true
        }
    }

    exclude(colorVal, colorName) {
        switch (colorName) {
            case "cmyk":
                return (colorVal.match(this.exclusionRegex.cmyk)) ? false : true
            case "hex3":
            case "hex4":
            case "hex6":
            case "hex8":
                return (colorVal.match(this.exclusionRegex.hex)) ? false : true
            case "grayscale":
                return (colorVal.match(this.exclusionRegex.grayscale)) ? false : true
            case "rgbDecimal":
                var colorLenght = colorVal.split(/[^0-9a-z]/)
                var indexOfDeciaml = colorVal.indexOf('rgb') > -1 || colorVal.indexOf('decimal') > -1
                return (colorLenght.length > 2 && !indexOfDeciaml) ? false : true  
            case "pantone": 
                return (colorVal.match(this.exclusionRegex.pantone)) ?  false : true   
            case "rgb":   
                return (colorVal.match(this.exclusionRegex.rgb)) ?  false : true     
            default:
                return false
        }
    }
}

class colorIdentify extends abstractColor {
    constructor(input) {
        super()
        this.input = input
        this.raw = new sanitizer(input)
    }

    get cmyk() {
        const { cmyk } = this.raw
        if (cmyk && this.exclude(this.input, 'cmyk')) {
            for (const i of 'cmyk') {
                cmyk[i] = this.abstractMakeInt(cmyk[i])
                if (isNaN(cmyk[i]) || cmyk[i] < 0 || cmyk[i] > 100) {
                    return false
                }
            }
            return true
        }

        return false
    }

    falseCmyk(stringCmyk) {
        let cmykArray = stringCmyk.replace(/[a-z]/g, '').split(/[^0-9]/)
        if (cmykArray.length === 4) {
            return true
        }
        return false
    }

    falseRgb(stringCmyk) {
        let cmykArray = stringCmyk.replace(/[a-z]/g, '').split(/[^0-9]/)
        if (cmykArray.length === 3) {
            return true
        }
        return false
    }

    get grayscale() {
        let { grayscale } = this.raw
        if (grayscale && this.exclude(this.input, 'grayscale')) {
            return grayscale >= 0 && grayscale <= 100
        }
        return false
    }

    get hex3() {
        const { hex3 } = this.raw
        return (this.exclude(this.input, 'hex3') && hex3.length === 3) ? hex3 : false
    }

    get hex4() {
        const { hex4 } = this.raw
        return (this.exclude(this.input, 'hex4') && hex4.length === 4) ? hex4 : false
    }

    get hex6() {
        const { hex6 } = this.raw
        return (this.exclude(this.input, 'hex3') && hex6.length === 6) ? hex6 : false
    }

    get hex8() {
        const { hex8 } = this.raw
        return (this.exclude(this.input, 'hex8') && hex8.length === 8) ? hex8 : false
    }

    get html() {
        const { html } = this.raw
        if (html) {
            return convertColor.pullDataFromList('html', html)
        }
        return false
    }

    get hsl() {
        const { hsl } = this.raw
        if (hsl) {
            for (const i of 'hsl') {
                hsl[i] = this.abstractMakeInt(hsl[i])
            }

            if (hsl.h >= 0 && hsl.h <= 360 && hsl.s >= 0 && hsl.s <= 100 && hsl.l >= 0 && hsl.l <= 100) {
                if (this.procentRequire(hsl.s, hsl.l)) {
                    hsl.s *= 100
                    hsl.l *= 100
                }
                return hsl
            }
        }

        return false
    }

    get hsv() {
        const { hsv } = this.raw
        if (hsv) {
            for (const i of 'hsv') {
                hsv[i] = this.abstractMakeInt(hsv[i])
            }

            if (hsv.h >= 0 && hsv.h <= 360 && hsv.s >= 0 && hsv.s <= 100 && hsv.v >= 0 && hsv.v <= 100) {
                if (this.procentRequire(hsv.s, hsv.l)) {
                    hsv.s *= 100
                    hsv.v *= 100
                }
                return hsv
            }
        }

        return false
    }

    get lab() {
        const { lab } = this.raw
        if (lab) {

            for (const i of 'lab') {
                lab[i] = this.abstractMakeInt(lab[i])
            }

            if (
                (lab.l >= 0 && lab.l <= 100) &&
                (lab.a >= -128 && lab.a <= 127) &&
                (lab.b >= -128 && lab.b <= 127)) {
                return lab
            }
        }

        return false
    }

    get pantone() {
        const { pantone } = this.raw
        if (pantone && this.exclude(this.input, 'pantone')) {
            if (pantone <= 5875 && pantone >= 100) {
                return pantone
            }
        }

        return false
    }

    get ral() {
        const { ral } = this.raw
        if (ral) {
            return convertColor.pullDataFromList('ral', ral)
        }
        return false
    }

    get rgb() {
        const { rgb } = this.raw
        if (rgb && this.exclude(this.input, 'rgb')) {
            for (const i of 'rgb') {
                rgb[i] = this.abstractMakeInt(rgb[i], 255)
                if (isNaN(rgb[i]) || rgb[i] < 0 || rgb[i] > 255) {
                    return false
                }
            }
            return rgb
        }

        return false
    }

    get rgba() {
        const { rgba } = this.raw

        if (rgba) {
            for (const i of 'rgb') {
                rgba[i] = this.abstractMakeInt(rgba[i], 255)
                if (isNaN(rgba[i]) || rgba[i] < 0 || rgba[i] > 255) {
                    return false
                }
            }
            rgba.a = parseFloat(rgba.a)
            return (rgba.a >= 0 && rgba.a <= 1) ? rgba : false

        }

        return false
    }

    get rgbDecimal() {
        const { rgbDecimal } = this.raw

        if (typeof rgbDecimal === 'string') {
            let numericData = rgbDecimal.match(/(\d+)/g)
            if (numericData) {
                numericData = numericData[0]

                if (rgbDecimal.indexOf('rgb') > -1 || rgbDecimal.indexOf('decimal') > -1) {
                    return numericData
                } else {
                    if (this.exclude(this.input, 'rgbDecimal') && this.abstractMakeInt(numericData) > 65792) {
                        return numericData
                    } else {
                        return false
                    }

                }
            }
        }

        return false
    }

    get w() {
        let { w } = this.raw
        if (w) {
            if (typeof w === 'string' && w.indexOf(w) > -1) {
                w = this.makeInt(w.replace(/[^0-9]/, ''))
            }
            if (typeof w === 'number') {
                return (w >= 380 && w <= 780) ? w : false
            }
        }

        return false
    }

    get xyz() {
        const { xyz } = this.raw
        if (xyz) {
            for (const i of 'xyz') {
                xyz[i] = this.abstractMakeInt(xyz[i])
                if (xyz[i] <= 0) {
                    return false
                }
            }
            return xyz
        }
        return false

    }

    get yuv() {
        const { yuv } = this.raw
        if (yuv) {
            for (const i of 'yuv') {
                yuv[i] = this.abstractMakeInt(yuv[i])
                if (yuv[i] < 0 || yuv[i] > 256) {
                    return false
                }
            }
            return yuv
        }
        return false
    }
}

module.exports = function (input, forceFormat) {


    const accepted_colors = new AcceptedColors();
    const identify = new colorIdentify(input)

    if (forceFormat) {
        return identify[forceFormat]
    } else {
        for (const aColor of accepted_colors.keys) {
            if (identify[aColor]) {
                return aColor
            }
        }
    }

    if (identify.falseCmyk(input)) {
        return 'cmyk'
    }

    if (identify.falseRgb(input)) {
        return 'rgb'
    }


    return false
}

components/colorScaffolding.js
'use strict'
const AcceptedColors = require('./accepted_colors')
const convertColor = require("./convert_color");

// ----------------------------- 0 | object color
class AbstractColorObject {
    constructor(data, colotType) {
        Object.assign(this, data)
        if (colotType) {
            Object.defineProperty(this, "colotType", {
                enumerable: false,
                writable: true,
                value: colotType
            });
        }
    }

    get hex() {
        return convertColor.convert({
            from: this.colotType,
            to: 'hex6',
            color: this,
            pretty: true,
        })
    }

    get clean() {
        let keys = Object.keys(this).join('')
        let values = Object.values(this).map(x => Math.round(x * 100) / 100).join(', ')
        return `${keys}(${values})`
    }
}

class AbstractColorRAL extends AbstractColorObject {
    constructor(data, colotType) {
        super(data)
    }

    get hex() {
        return convertColor.convert({
            from: 'ral',
            to: 'hex6',
            color: this.ral.toString(),
            pretty: true,
        })
    }

    get clean() {
        const { ral } = this
        return `RAL ${ral}`
    }
}

// ----------------------------- 1 | string color
class AbstractColorString extends String {
    constructor(data, colotType) {
        super(data)
        if (colotType) {
            Object.defineProperty(this, "colotType", {
                enumerable: false,
                writable: true,
                value: colotType
            });
        }
    }

    get hex() {
        return convertColor.convert({
            from: this.colotType,
            to: 'hex6',
            color: this.toString(),
            pretty: true,
        })
    }

    get clean() {
        let _this = this.toString().replace(/([A-Z])/g, ' $1').replace(/ {2}|   {2}/g, ' ').trim()
        return _this.charAt(0).toUpperCase() + _this.slice(1)
    }
}

class AbstractColorStringRAL extends String {
    constructor(data, colotType) {
        super(data)
        if (colotType) {
            Object.defineProperty(this, "colotType", {
                enumerable: false,
                writable: true,
                value: colotType
            });
        }
    }

    get hex() {
        return convertColor.convert({
            from: this.colotType,
            to: 'hex6',
            color: this.toString(),
            pretty: true,
        })
    }

    get clean() {
        return `RAL ${this}`
    }
}

class AbstractColorHTML extends AbstractColorString {
    constructor(data, colotType) {
        super(data)
    }

    get hex() {
        return convertColor.convert({
            from: 'html',
            to: 'hex6',
            color: this.toString(),
            pretty: true,
        })
    }
}

class AbstractColorHex extends AbstractColorString {
    constructor(data, colotType) {
        super(data)
    }

    get hex() {
        return convertColor.convert({
            from: 'hex' + this.length,
            to: 'hex6',
            color: this.toString(),
            pretty: true,
        })
    }

    get clean() {
        return `#${this}`
    }
}

// ----------------------------- 2 | numeric color
class AbstractColorNumber extends Number {
    constructor(data, colotType) {
        super(data)
    }
}

class AbstractColorPantone extends AbstractColorNumber {
    constructor(data, colotType) {
        super(data)
    }

    get hex() {
        return convertColor.convert({
            from: 'pantone',
            to: 'hex6',
            color: this.toString(),
            pretty: true,
        })
    }

    get clean() {
        return `Pantone ${this}`
    }
}

class AbstractColorGrayscale extends AbstractColorNumber {
    constructor(data, colotType) {
        super(data)
    }

    get hex() {
        return convertColor.convert({
            from: 'grayscale',
            to: 'hex6',
            color: this,
            pretty: true,
        })
    }

    get clean() {
        return `Grayscale: ${this}`
    }
}

module.exports = {
    clone: function (data) {
        if (typeof data === "object") {
            return JSON.parse(JSON.stringify(data))
        } else {
            return data
        }
    },

    segments: function({startPoint,endpoint, segments, notIncludeStart = false, notIncludeEnd = false}) {
        const totalSize = endpoint - startPoint
        const segmentsArray = []
        const segmentSize = totalSize/segments


        for(let i = (notIncludeStart)? 1: 0 ; i <= ((notIncludeEnd)? segments -1: segments); i++) {
            segmentsArray.push((segmentSize*i) + startPoint)
        }
        
        return segmentsArray
    },

    hslCircle: function (value){
        if(value < 0){
          return Math.abs(360+value)%360
        } else {
          return (value)%360
        }
    },

    hslSegment: function ({ hue, segments, offset = 0 }) {
        const arcSize = (360 / segments)
        return this.hslCircle((Math.round((hue - offset) / arcSize ) * arcSize ) + offset)
    },

    customColorObject: function ({ colorData, typeOfColor }) {
        if (typeof colorData === 'string') {
            if (typeOfColor === 'html') {
                return new AbstractColorHTML(colorData, typeOfColor)
            } else if (typeOfColor.indexOf('hex') > -1) {
                return new AbstractColorHex(colorData, typeOfColor)
            } else if (typeOfColor === 'ral') {
                return new AbstractColorStringRAL(colorData, typeOfColor)
            }
            return new AbstractColorString(colorData, typeOfColor)
        } else if (typeof colorData === 'number') {
            if (typeOfColor === 'pantone') {
                return new AbstractColorPantone(colorData, typeOfColor)
            } else if (typeOfColor === 'grayscale') {
                return new AbstractColorGrayscale(colorData, typeOfColor)
            }
        } else if (typeof colorData === 'object') {
            if (typeOfColor === 'ral') {

                return new AbstractColorRAL(colorData, typeOfColor)
            }
            return new AbstractColorObject(colorData, typeOfColor)
        } else {
            return colorData
        }
    },

    get acceptedColors() {
        return new AcceptedColors()
    }
}

dot-color.js
'use strict'
/* eslint-disable */
const sanitize = require('./components/sanitization')
const identify = require('./components/identify')
const compare = require('./components/compere.js')
const convertColor = require('./components/convert_color')
const colorScaffolding = require('./components/colorScaffolding')

// ----------------------------- A | Object Class
class objectiveColor {
    constructor(inputColor) {
        this.init()
        this.colorExtractor(inputColor)
    }

    colorExtractor(inputColor, setFormat) {
        delete this.format
        delete this.sanitizedColor
        if (inputColor) {
            this.format = (setFormat) ? ((identify(inputColor, setFormat)) ? setFormat : false) : identify(inputColor)
            
            if (this.format) {
                this.sanitizedColor = sanitize(inputColor, this.format)
            } else {
                this.format = false
                this.sanitizedColor = false
            }
        }
    }

    init() {
      for (const colorType of this.acceptedColors.keys) {
        Object.defineProperty(this, colorType, {
          get() {
            let tempColor = false;
            let _this = this;

            if (this.format === colorType) {
              tempColor = this.sanitizedColor;
            } else if (this.format) {
              tempColor = convertColor.convert({
                from: this.format,
                to: colorType,
                color: this.sanitizedColor,
              });
            }

            return this.customColorObject({
              colorData: tempColor,
              typeOfColor: colorType,
            });
          },
          set(input) {
            this.colorExtractor(input, colorType);
          },
        });
      }

      // ----------- Alias
      Object.defineProperty(this, "color", {
        get() {
          return this.sanitizedColor;
        },
        set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "hex", {
        get() {
          return this.hex6;
        },
        set(input) {
          this.colorExtractor(input);
        },
      });

      // ----------- converts
      Object.defineProperty(this, "invert", {
        get() {
          const { hsl } = this;
          hsl.h = this.hslCircle(hsl.h + 180)

          const tempColor = convertColor.convert({
            from: "hsl",
            to: this.format,
            color: hsl,
          });
          return this.customColorObject({
            colorData: tempColor,
            typeOfColor: this.format,
          });
        },
        set(input) {
          this.colorExtractor(input);
        },
      });
      
      Object.defineProperty(this, "primary", {
        get() {
          const tempH = this.hslSegment({ hue: this.hsl.h, segments: 3 });
          const tempColor = convertColor.convert({
            from: "hsl",
            to: this.format,
            color: { h: tempH, s: 100, l: 50 },
          });
          return this.customColorObject({
            colorData: tempColor,
            typeOfColor: this.format,
          });
        },
        set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "secondary", {
        get() {
          const tempH = this.hslSegment({
            hue: this.hsl.h,
            segments: 3,
            offset: 60,
          });
          const tempColor = convertColor.convert({
            from: "hsl",
            to: this.format,
            color: { h: tempH, s: 100, l: 50 },
          });
          return this.customColorObject({
            colorData: tempColor,
            typeOfColor: this.format,
          });
        },
        set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "tertiary", {
        get() {
          const tempH = this.hslSegment({
            hue: this.hsl.h,
            segments: 6,
            offset: 30,
          });
          const tempColor = convertColor.convert({
            from: "hsl",
            to: this.format,
            color: { h: tempH, s: 100, l: 50 },
          });
          return this.customColorObject({
            colorData: tempColor,
            typeOfColor: this.format,
          });
        },
        set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "tones", {
        get() {
          const { hsl, format } = this

          const _this = {
            hslArray: [],
            hslSegments: this.segments({startPoint: 0, endpoint: 100, segments: 10}),
            buildArray: function () {
              for(let i = 0; i < this.hslSegments.length; i++ ) {
                const colorTemp = convertColor.convert({
                  from: "hsl",
                  to: format,
                  color: {h: hsl.h, s:hsl.s, l:this.hslSegments[i] },
                });
                this.hslArray.push(colorTemp)
              }
            }
          }
          _this.buildArray()
          return _this.hslArray
        },set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "tints", {
        get() {
          const { hsl, format } = this

          const _this = {
            hslArray: [],
            hslSegments: this.segments({startPoint: hsl.l, endpoint: 100, segments: 10}),
            buildArray: function () {
              for(let i = 0; i < this.hslSegments.length; i++ ) {
                const colorTemp = convertColor.convert({
                  from: "hsl",
                  to: format,
                  color: {h: hsl.h, s:hsl.s, l:this.hslSegments[i] },
                });
                this.hslArray.push(colorTemp)
              }
            }
          }
          _this.buildArray()
          return _this.hslArray
        },set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "shades", {
        get() {
          const { hsl, format } = this

          const _this = {
            hslArray: [],
            hslSegments: this.segments({startPoint: 0, endpoint: hsl.l, segments: 10}),
            buildArray: function () {
              for(let i = 0; i < this.hslSegments.length; i++ ) {
                const colorTemp = convertColor.convert({
                  from: "hsl",
                  to: format,
                  color: {h: hsl.h, s:hsl.s, l:this.hslSegments[i] },
                });
                this.hslArray.push(colorTemp)
              }
            }
          }
          _this.buildArray()
          return _this.hslArray
        },set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "analogous", {
        get() {
          const { hsl, format } = this
          const _this = {
            hslArray: [],
            hslSegments: [this.hslCircle(hsl.h-30), hsl.h, this.hslCircle(hsl.h+30) ],
            buildArray: function () {
              for(let i = 0; i < this.hslSegments.length; i++ ) {
                const colorTemp = convertColor.convert({
                  from: "hsl",
                  to: format,
                  color: {h: this.hslSegments[i], s:hsl.s, l:hsl.l },
                });

                this.hslArray.push(colorTemp)
              }
            }
          }

          _this.buildArray()
          return _this.hslArray
        },set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "complementary", {
        get() {
          const a = this.invert.clean
          const b = this.customColorObject({
            colorData: this.sanitizedColor,
            typeOfColor: this.format,
          }).clean
          return [a,b]
        },set(input) {
          this.colorExtractor(input);
        },
      });

      Object.defineProperty(this, "warmth", {
        get() {
          let _this = {
            isWarm: true,
            warm: 0,
            cold: 0
          }

          let { h } = this.hsl
          h = h - 60 
          if(h < 0) {
            h = 360 - h
          }

          if( h > 180){
            _this.isWarm = false
            _this.cold = 180 / (h%180)
            _this.warm = 180/ (Math.abs(180 - (h%180)))
          } else  {
            _this.cold = Math.abs(180 - (h%180))
            _this.warm = h%180
          }

          return _this
        },set(input) {
          this.colorExtractor(input);
        },
      });
    }

    compare (...arrayOfColor) {
      let _this = []
      for( const i of arrayOfColor ) {
        let temp = { raw: i }

        temp.identify = identify(temp.raw)
        temp.raw = sanitize(temp.raw, temp.identify) 

        temp.rgb = convertColor.convert({
          from: temp.identify,
          to: "rgb",
          color: temp.raw,
        })

        temp.hex = convertColor.convert({
          from: "rgb",
          to: "hex6",
          color: temp.rgb,
        })

        temp.compare = compare.init({ c1: temp.rgb , c2: this.rgb })

        _this.push(temp)
      }

      _this.sort(function(a, b) {
        if (a.compare < b.compare) { return -1 }
        if (a.compare > b.compare) { return 1 }
        return 0;
      });

      let _simplified = []
      for( const i of _this ) {
        _simplified.push(i.raw)
      }

      return _simplified
  }
}

Object.assign(objectiveColor.prototype, colorScaffolding);

module.exports = objectiveColor